cmd/gateway/azure/gateway-azure.go

```diff
@@ -144,50 +145,71 @@ func (g *Azure) Name() string {
 	return azureBackend
 }
 
-// All known cloud environments of Azure
-var azureEnvs = []azure.Environment{
-	azure.PublicCloud,
-	azure.USGovernmentCloud,
-	azure.ChinaCloud,
-	azure.GermanCloud,
-}
-
 // NewGatewayLayer initializes azure blob storage client and returns AzureObjects.
 func (g *Azure) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error) {
-	var err error
-	// The default endpoint is the public cloud
-	var endpoint = azure.PublicCloud.StorageEndpointSuffix
-	var secure = true
+	var endpoint string
 
 	// Load the endpoint url if supplied by the user.
 	if g.host != "" {
-		endpoint, secure, err = minio.ParseGatewayEndpoint(g.host)
+		host, secure, err := minio.ParseGatewayEndpoint(g.host)
 		if err != nil {
 			return nil, err
 		}
-		// Reformat the full account storage endpoint to the base format.
-		//   e.g. testazure.blob.core.windows.net => core.windows.net
-		endpoint = strings.ToLower(endpoint)
-		for _, env := range azureEnvs {
-			if strings.Contains(endpoint, env.StorageEndpointSuffix) {
-				endpoint = env.StorageEndpointSuffix
-				break
-			}
+
+		var protocol string
+		if secure {
+			protocol = "https"
+		} else {
+			protocol = "http"
 		}
+
+		// for containerized storage deployments like Azurite or IoT Edge Storage,
+		// account resolution isn't handled via a hostname prefix like
+		// `http://${account}.host/${path}` but instead via a route prefix like
+		// `http://host/${account}/${path}` so adjusting for that here
+		if !strings.HasPrefix(host, fmt.Sprintf("%s.", creds.AccessKey)) {
+			host = fmt.Sprintf("%s/%s", host, creds.AccessKey)
+		}
+
+		endpoint = fmt.Sprintf("%s://%s", protocol, host)
+	} else {
+		endpoint = fmt.Sprintf("https://%s.blob.core.windows.net", creds.AccessKey)
```

Comment:
It looks like with this change Gov cloud (and other non std cloud) is not supported anymore, any reason for that ?

Reply:
If I understood the old code correctly, this change shouldn't affect the types of storage accounts we can handle and the code change should support any blob endpoint:

- If no custom host is provided via the command line then we assume that the storage account is in Azure public and set the endpoint URL accordingly. For example, `minio gateway azure` will use the endpoint URL `https://${accountName}.blob.core.windows.net`.

- If a custom host is provided, it gets passed verbatim to the underlying SDK which provides support for GovCloud. For example, `minio gateway azure ${accountName}.blob.core.usgovcloudapi.net` will use the endpoint URL `https://${accountName}.blob.core.usgovcloudapi.net`.

- If the custom host doesn't start with the account name, then we're targeting a containerized storage account like Azurite and have to do some post-processing to ensure that the account route prefix gets handled correctly. For example, `minio gateway azure http://localhost:10000` will use the endpoint URL `http://localhost:1000/${accountName}`.

I've added some tests in 3afdfec to make the intent of the code clearer (will squash/rebase before merge). Please let me know if you're okay with the described behavior or if I misunderstood something.